testutil/network.go

```diff
@@ -0,0 +1,358 @@
+		appCfg.MinGasPrices = cfg.MinGasPrices
+		appCfg.API.Enable = true
+		appCfg.API.Swagger = false
+		appCfg.Telemetry.Enabled = false
+
+		ctx := server.NewDefaultContext()
+		tmCfg := ctx.Config
+		tmCfg.Consensus.TimeoutCommit = cfg.TimeoutCommit
+
+		// Only allow the first validator to expose an RPC and API server/client
+		// due to Tendermint in-process constraints.
+		apiAddr := ""
+		tmCfg.RPC.ListenAddress = ""
+		if i == 0 {
+			apiListenAddr, _, err := server.FreeTCPAddr()
+			require.NoError(t, err)
+			appCfg.API.Address = apiListenAddr
+
+			apiURL, err := url.Parse(apiListenAddr)
+			require.NoError(t, err)
+			apiAddr = fmt.Sprintf("http://%s:%s", apiURL.Hostname(), apiURL.Port())
+
+			rpcAddr, _, err := server.FreeTCPAddr()
+			require.NoError(t, err)
+			tmCfg.RPC.ListenAddress = rpcAddr
+		}
+
+		logger := log.NewNopLogger()
+		if cfg.EnableLogging {
+			logger = log.NewTMLogger(log.NewSyncWriter(os.Stdout))
+			logger, _ = tmflags.ParseLogLevel("info", logger, tmcfg.DefaultLogLevel())
+		}
+
+		ctx.Logger = logger
+
+		nodeDirName := fmt.Sprintf("node%d", i)
+		nodeDir := filepath.Join(network.BaseDir, nodeDirName, "simd")
+		clientDir := filepath.Join(network.BaseDir, nodeDirName, "simcli")
+		gentxsDir := filepath.Join(network.BaseDir, "gentxs")
+
+		require.NoError(t, os.MkdirAll(filepath.Join(nodeDir, "config"), 0755))
+		require.NoError(t, os.MkdirAll(clientDir, 0755))
+
+		tmCfg.SetRoot(nodeDir)
+		tmCfg.Moniker = nodeDirName
+		monikers[i] = nodeDirName
+
+		proxyAddr, _, err := server.FreeTCPAddr()
+		require.NoError(t, err)
+		tmCfg.ProxyApp = proxyAddr
+
+		p2pAddr, _, err := server.FreeTCPAddr()
+		require.NoError(t, err)
+		tmCfg.P2P.ListenAddress = p2pAddr
+		tmCfg.P2P.AddrBookStrict = false
+		tmCfg.P2P.AllowDuplicateIP = true
+
+		nodeID, pubKey, err := genutil.InitializeNodeValidatorFiles(tmCfg)
+		require.NoError(t, err)
+		nodeIDs[i] = nodeID
+		valPubKeys[i] = pubKey
+
+		kb, err := keyring.New(sdk.KeyringServiceName(), keyring.BackendTest, clientDir, buf)
+		require.NoError(t, err)
+
+		addr, secret, err := server.GenerateSaveCoinKey(kb, nodeDirName, cfg.Passphrase, true)
+		require.NoError(t, err)
+
+		info := map[string]string{"secret": secret}
+		infoBz, err := json.Marshal(info)
+		require.NoError(t, err)
+
+		// save private key seed words
+		require.NoError(t, writeFile(fmt.Sprintf("%v.json", "key_seed"), clientDir, infoBz))
+
+		balances := sdk.NewCoins(
+			sdk.NewCoin(fmt.Sprintf("%stoken", nodeDirName), cfg.AccountTokens),
+			sdk.NewCoin(cfg.BondDenom, cfg.StakingTokens),
+		)
+
+		genFiles = append(genFiles, tmCfg.GenesisFile())
+		genBalances = append(genBalances, banktypes.Balance{Address: addr, Coins: balances.Sort()})
+		genAccounts = append(genAccounts, authtypes.NewBaseAccount(addr, nil, 0, 0))
+
+		createValMsg := stakingtypes.NewMsgCreateValidator(
+			sdk.ValAddress(addr),
+			valPubKeys[i],
+			sdk.NewCoin(sdk.DefaultBondDenom, cfg.BondedTokens),
```

Comment:
Perhaps an extension worth implementing in a future PR. In addition to different validator configs as mentioned above, we should allow initializing networks with unequal token distributions

Reply:
Absolutely!